pup-0003.md

```diff
@@ -0,0 +1,116 @@
+create a barrier to entry for community members. With the cherry-pick model,
+all PRs are opened against master and the onus to update release branches is on
+the person merging (who usually has better knowledge of which branches a commit
+needs to be applied to).
+
+Another minor caveat is that changes that we want to only be made to older
+release branches (but not to newer ones) still need to be merged forward. We
+could elminate the need for this with cherry picking. It's worth noting that
+this rarely happens.
+
+Lastly, we've also seen some mistakes where branches are accidentally merged
+into other branches. Commits could be accidentally cherry picked but this would
+be less disastrous since it's only a single commit.
+
+
+## Detailed Design
+
+Here is an overview of the workflow:
+
+1. All PRs would be opened against master regardless of whether they contain a
+   feature, bug fix, or hotfix.
+2. After the PR is ready to be merged, the person merging the commit would
+   first merge the commit to master and then cherry pick the commit to any
+   release branches that required the change. For a bugfix, this would be the
+   latest release branch. For hotfixes, this would be the oldest affected
+   release branch plus any newer branches.
+3. The person merging would also need to track in redmine which release
+   branches the change has been made to.
+
+### Updating redmine
+
+For the third step, a few proposals for how to track this information in
+redmine include:
+
+* Use a second/different redmine issue to track the inclusion of the backport
+into a given release and associate the two somehow. Maybe a custom field called
+'cherry picks' could record issue numbers as a comma separated list which would
+be clickable into a redmine query.
+* Use an associated commit. With this a redmine issue will show the original
+commits and any cherry picks. We could have a new commit keyword so a commit
+could contain "cherrypick #1234" which would associate the cherry pick commit
+with issue 1234. It would not modify the state of the issue so whatever state
+it's in, e.g. CLOSED - CURRENT RELEASE or MODIFIED would remain as-is.
+* Use a custom field to track the cherry picks. This is convenient for release
+engineering queries, but would not show the actual cherry pick commits
+themselves.
+
+
+## Drawbacks
+
+Git is really built for changes to be merged between branches. For example,
+when cherry picking a change, the commit is assigned a new id but when merging
+the commit, it retains the same id. It makes looking up the commit to see what
+branches it's been applied to easier. If we cherry pick, we have to track this
+information externally to git. This means also having to remember to update
+redmine.
+
+Switching to this model also creates some inconsistency. For example, if we
+adopt this for Pulp 3, then we still have to use merge forwarding for Pulp 2.
+Then we are stuck having to support two different strategies.
+
```

Comment:
Defaulting to master, and requiring extra steps to get a change onto the current release branch, is likely to result in fewer bug fixes making it onto current release branches. We can and should try to be diligent about cherry-picking fixes to the release branch, but the path of least resistance may prove too tempting at times.